src/DocumentDB.ChangeFeedProcessor/PartitionManagement/ILease.cs

```diff
@@ -49,6 +49,11 @@ public interface ILease
         /// </summary>
         string ConcurrencyToken { get; }
 
+        /// <summary>
+        /// Gets the timestamp of the lease update on the server.
+        /// </summary>
+        DateTime ServerTimestamp { get; }
```

Comment:
Since this is not needed for public, let's not add it. It's fine to have it as private in the impl class.

Reply:
The problem is that in case of time skew, it would be good to detect it (at least to log it, in the load balancer). 

Imagine that the time on instance and docdb server diverge. We need to be able to recover from it => so let's start with detecting. As we use ILease, we dont know the server time.

Reply:
I don't like different timestamps on ILease as it may get messy. Ideally all diff between local and server should be hidden by underlying layer, at least by CFP library. We should probably revisit this.

It looks like today we persist (not sure why) ExplicitTimestamp (which is v2 only concept and is set every time we update the lease) to the server so that it will be taken by another host and thus may diverge even more (client-to-client rather than client-to-server). In v1 we have logic to account for time diff between local and server which kind of worked -- the load balancer when determining if lease is expired used adjusted local vs server time.

Reply:
I had similar mixed feelings. Just to accelerate the issue. We had a production issue with other service which was caused by time's skew in the production by 1 instance. 

So using local timestamp requires that ALL processor instances will be synced, e.g. we have a script running periodically on each instance in very DC which syncs the time against closest NTP server.

All in all, the best would be to move this responsibility to DocDB server. Could we  rely on DocDB servers to be time synced with a error precisions <1s?  

If so, I would leverage server time only. Proposal:
1) when we finish querying the leases, we could use a time from the server. In the georeplicated scenario it would required DocDB servers to be sync. Can we rely on it? 
e.g. when doing query.ExecuteAsync() we would get the server time from the header of the first response:
Date: Fri, 07 Sep 2018 09:08:39 GMT

and then we would compare it against _ts on leases.

2) using ONLY _ts -> on every load balancing strategy run it would update 1 document and get the _ts from the server master replica and use it for a comparison. 

wdyt?

Reply:
agreeed! no2 with periodic check.

Reply:
Also, forgot to mention... A better way would be to use TTL-expiration for leases, this would be server-only logic. Problem is that lease doc on expire will be deleted along with data that should stay on expiration. This can be solved by splitting the lease into 2 docs, one of them would expire, the other one will not. We discussed that in v1 timeframe but took simpler/cheaper approach of adjusted _ts. I think eventually we should move to use TTL, no.2 can be short term solution.

Reply:
:) same here

we combine 2 different things together and we could leverage TTL for it. 

but the hardest thing now is how to migrate from old to new model w/o downtime. What we could do is to have 2 modes for quite some time and hope that people move to the new library version which would support both version (would be storing data in old and new way). After some time, only new version would be used. or do u see any other idea?

Reply:
but that would be for different PR. 

so let's get this PR done - so let's change then ILease so that it will expose ServerTimestamp aka heartbeat timestamp and let's log it. 

when having 2 documents (as discussed previously), we would combine both documents under ILease, wdyt?

Reply:
OK (though I would prefer different interface or downcase), let's keep it but please add a comment something like this property is temporary and is only used for logging internally in CFP, essentially, don't make dependency on it. In subsequent change we will address timeouts and remove this.

Reply:
so I used it only in logging and as discussed, we will address it later in specific PR focused on lease documents expiration,